src/main/java/io/cdap/plugin/gcp/publisher/source/PubSubDirectDStream.java

```diff
@@ -0,0 +1,292 @@
+    this.config = config;
+    this.readDuration = readDuration;
+    this.context = context;
+    this.autoAcknowledge = autoAcknowledge;
+    this.mappingFn = mappingFn;
+    this.pipeline = context.getPipelineName();
+    this.credentials = createCredentials(config.getServiceAccount(), config.isServiceAccountFilePath());
+  }
+
+  @Override
+  public Option<RDD<T>> compute(Time validTime) {
+    LOG.debug("Computing RDD for time {}.", validTime);
+    PubSubRDD pubSubRDD = new PubSubRDD(streamingContext.sparkContext(), validTime, readDuration, config,
+                                        autoAcknowledge, credentials);
+    RDD<T> mapped = pubSubRDD.map(mappingFn, scala.reflect.ClassTag$.MODULE$.apply(PubSubMessage.class));
+    return Option.apply(mapped);
+  }
+
+  @Override
+  public void start() {
+    if (config.getTopic() != null) {
+      try {
+        createSubscriptionIfNotPresent();
+      } catch (IOException e) {
+        throw new RuntimeException("Subscription creation failed.", e);
+      } catch (InterruptedException e) {
+        throw new RuntimeException("Subscription creation failed.", e);
+      }
+    }
+
+    try {
+      subscriptionAdminClient = buildSubscriptionAdminClient(credentials);
+    } catch (IOException e) {
+      throw new RuntimeException("SubscriptionAdminClient creation failed.", e);
+    }
+
+    // If state with snapshot is present, seek to it.
+    this.currentSnapshotName = fetchSnapShot(config.getSubscription(), context);
+    if (this.currentSnapshotName != null) {
+      seekSnapshot(currentSnapshotName, ProjectSubscriptionName.of(config.getProject(), config.getSubscription()));
+      // Can use the same snapshot till end of the batch.
+      takeSnapshot = false;
+    } else {
+      takeSnapshot = true;
+    }
+  }
+
+  @Override
+  public void stop() {
+    if (subscriptionAdminClient == null) {
+      return;
+    }
+
+    subscriptionAdminClient.close();
+  }
+
+  @Override
+  public void onBatchStarted(io.cdap.cdap.etl.api.streaming.StreamingContext streamingContext) {
+    LOG.debug("Starting a batch.");
+    if (!takeSnapshot) {
+      return;
+    }
+
+    String generatedSnapshotName = generateName(config.getSubscription());
+    ProjectSnapshotName projectSnapshotName = ProjectSnapshotName.of(config.getProject(), generatedSnapshotName);
+    ProjectSubscriptionName projectSubscriptionName = ProjectSubscriptionName.of(config.getProject(),
+                                                                                 config.getSubscription());
+    // Create in Pub/Sub
+    Snapshot snapshot = createSnapshot(projectSnapshotName, projectSubscriptionName);
+    // Save snapshot as state
+    try {
+      saveSnapshotAsState(snapshot, config.getSubscription(), context);
+    } catch (IOException e) {
+      // Delete the snapshot from Pub/Sub to avoid abandoned snapshots.
+      deleteSnapshot(projectSnapshotName);
```

Comment:
though unlikely, this can fail too. Is there any concept of a TTL for snapshots?